crates/topos-tce-api/src/runtime/task.rs

```diff
@@ -0,0 +1,173 @@
+use crate::stream::StreamCommand;
+use futures::stream::FuturesUnordered;
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::future::{Future, IntoFuture};
+use std::pin::Pin;
+use tokio::sync::mpsc::Sender;
+use topos_api::grpc::checkpoints::TargetStreamPosition;
+use topos_core::types::stream::CertificateTargetStreamPosition;
+use topos_core::types::CertificateDelivered;
+use topos_core::uci::SubnetId;
+use topos_tce_storage::{FetchCertificatesFilter, FetchCertificatesPosition, StorageClient};
+use tracing::{error, info};
+use uuid::Uuid;
+
+type TargetSubnetStreamPositions = HashMap<SubnetId, HashMap<SubnetId, TargetStreamPosition>>;
+pub(crate) type RunningTasks = FuturesUnordered<Pin<Box<dyn Future<Output = TaskStatus> + Send>>>;
+
+/// Status of a sync task
+///
+/// When registering a stream, a sync task is started to fetch certificates from the storage
+/// and push them to the stream.
+pub(crate) enum TaskStatus {
+    ///  The sync task is active and started running
+    Running,
+    /// The sync task failed and reported an error
+    Error,
+    /// The sync task exited gracefully and is done pushing certificates to the stream
+    Done,
+    /// The sync task was cancelled externally
+    Cancelled,
+}
+
+pub(crate) struct Task {
+    pub(crate) status: TaskStatus,
+    pub(crate) stream_id: Uuid,
+    pub(crate) target_subnet_stream_positions: TargetSubnetStreamPositions,
+    pub(crate) storage: StorageClient,
+    pub(crate) notifier: Sender<StreamCommand>,
+    pub(crate) exit_signal: tokio::sync::oneshot::Receiver<()>,
+}
+
+impl Task {
+    pub(crate) fn new(
+        stream_id: Uuid,
+        target_subnet_stream_positions: TargetSubnetStreamPositions,
+        storage: StorageClient,
+        notifier: Sender<StreamCommand>,
+    ) -> (tokio::sync::oneshot::Sender<()>, Self) {
```

Comment:
`new` function usually returns `Self`, if you pass the `exit_receiver` as an argument it could make the function following Rust guideline.